src/ir/auth_logic/souffle_emitter.h

```diff
@@ -0,0 +1,143 @@
+ * limitations under the License.
+ */
+
+#ifndef SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+#define SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+
+#include <unordered_set>
+
+#include "src/ir/auth_logic/ast.h"
+
+namespace raksha::ir::auth_logic {
+
+class SouffleEmitter {
+  public:
+    std::string EmitProgram(DLIRProgram program) {
+      SouffleEmitter emitter;
+      std::string body = emitter.EmitProgramBody(program);
+      std::string outputs = emitter.EmitOutputs(program);
+      std::string decls = emitter.EmitDeclarations();
+      return std::move(body) 
+        + "\n" 
+        + std::move(outputs) 
+        + "\n" 
+        + std::move(decls);
+    }
+
+  private:
+    SouffleEmitter() 
+      : declarations_(
+          std::unordered_set<Predicate, Predicate::HashFunction>()) 
+    {};
+
+    // This function produces a normalized version of predicates to
+    // be used when generating declarations of the predicates. It replaces
+    // argument names with generic numbered ones. It is applied
+    // to predicates before they are added to the set of declarations so that
+    // there are no duplicate delcarations (which would otherwise happen
+    // whenever a predicate is referenced more than once with different
+    // arguments).
+    // To prevent instances of negated and non-negated uses of the predicate 
+    // from generating two declarations, the sign here is always positive.
+    Predicate PredToDeclaration(Predicate predicate) {
+      std::vector<std::string> decl_args;
+      for( int i=0; i < predicate.args().size(); i++ ) {
+        decl_args.push_back("x" + std::to_string(i));
+      }
+      return Predicate(predicate.name(), predicate.args(), kPositive);
+    }
+
+    std::string EmitVecSeparated(std::vector<std::string> vec,
+        std::string sep) {
+      std::string ret = "";
+      bool is_first = true;
+      for(auto elm : vec) {
+        if( !is_first ) { ret += sep; } else { is_first = false; }
+        ret += elm;
+      }
+      return ret;
+    }
+
+    std::string EmitPredicate(Predicate predicate) {
+      // Whenever a new predicate is encountered, it is added to the set of 
+      // declarations (which does not include duplicates because it is a set).
+      declarations_.insert(PredToDeclaration(predicate));
+      return (predicate.sign() == kNegated ? "!" : "")
+          + predicate.name()
+          + "(" + EmitVecSeparated(predicate.args(), ", ") + ")";
+    }
+
+    std::string EmitAssertionInner(Predicate predicate) {
+      return EmitPredicate(predicate) + ".";
+    }
+
+    std::string EmitAssertionInner(DLIRCondAssertion cond_assertion) {
+      std::string lhs = EmitPredicate(cond_assertion.lhs());
+      std::vector<std::string> rhs_translated;
+      for(auto arg: cond_assertion.rhs()) {
+        rhs_translated.push_back(EmitPredicate(arg));
+      }
+      std::string rhs = EmitVecSeparated(rhs_translated, ", ");
+      return std::move(lhs) + " :- " + std::move(rhs) + ".";
```

Comment:
When combining multiple strings together like this, just use `absl::StrCat`. https://abseil.io/docs/cpp/guides/strings#abslstrcat